flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -291,13 +300,23 @@ public InputStatus emitNext(DataOutput<OUT> output) throws Exception {
 
         // short circuit the common case (every invocation except the first)
         if (currentMainOutput != null) {
-            return sourceReader.pollNext(currentMainOutput);
+            return pollNext();
         }
 
         // this creates a batch or streaming output based on the runtime mode
-        currentMainOutput = eventTimeLogic.createMainOutput(output);
+        currentMainOutput =
+                eventTimeLogic.createMainOutput(
+                        new MetricTrackingOutput<>(output, sourceMetricGroup));
         lastInvokedOutput = output;
-        return sourceReader.pollNext(currentMainOutput);
+        return pollNext();
+    }
+
+    private InputStatus pollNext() throws Exception {
+        InputStatus inputStatus = sourceReader.pollNext(currentMainOutput);
+        if (inputStatus == InputStatus.NOTHING_AVAILABLE) {
+            sourceMetricGroup.idlingStarted();
+        }
+        return inputStatus;
```

Comment:
Does this logic require a test?

Reply:
Yes, I didn't want to add tests until the design is approved to save some time. The proper PR on top of the FLIP needs pretty much a test per metric.